pmd-apex/src/test/java/net/sourceforge/pmd/lang/apex/ast/ApexLexerTest.java

```diff
@@ -49,4 +53,35 @@ void testParser() {
         ApexParser.CompilationUnitContext compilationUnit = parser.compilationUnit();
         assertNotNull(compilationUnit);
     }
+
+    @Test
+    void testLexerUnicodeEscapes() {
+        String s = "'Fran\\u00E7ois'";
+        assertEquals(2, getLexingErrors(CharStreams.fromString(s)));
+        assertEquals(0, getLexingErrors(new CaseInsensitiveInputStream(CharStreams.fromString(s))));
```

Comment:
Note, that this test doesn't test PMD - it is testing the underlying library. But it doesn't make sure, that PMD is using the underlying library correctly...
This means, if your fix is removed, this test stays green, although PMD will be broken...

Please add a test case, e.g. in https://github.com/pmd/pmd/blob/main/pmd-apex/src/test/java/net/sourceforge/pmd/lang/apex/ast/ApexCommentTest.java

The test could look something like this:

```java
    @Test
    void fileWithUnicodeEscapes() {
        ASTApexFile file = apex.parse(FORMAL_COMMENT_CONTENT + "\n"
                + "class MyClass { String s = 'Fran\\u00E7ois'; }");
        ASTFormalComment comment = file.descendants(ASTUserClass.class).children(ASTFormalComment.class).first();
        assertEquals(FORMAL_COMMENT_CONTENT, comment.getImage());
    }
```

Thanks to the custom error listener in ApexCommentBuilder, this should fail, if we don't apply your fix (using CaseInsensitiveInputStream) and should be working with your fix.

Reply:
Good points!

I agree, my original test doesn't cover the specific situation on how the library is being used by PMD. I intended it more as a showcase of the peculiar behavior of the apex-lexer library rather than as a proper unit test for `ApexCommentBuilder`. Because it's odd that the library apparently breaks on Unicode escape sequences in *strings* - those shouldn't be affected at all by the fact that the Apex *language* is case-insensitive. I will notify the maintainers of apex-parser because it may be a bug. My intention for the PR here was to raise awareness and provide a quick fix for PMD's use case while the underlying lexing issue is under investigation.

I noticed that `ApexCommentBuilder` doesn't have unit tests yet and as it's fairly complex and I'm not familiar with the codebase, writing a proper unit test seemed like a daunting task. It's indeed a good idea to add it `ApexCommentTest`.

Should I remove my original test?

Reply:
> Should I remove my original test?

I think, we can keep these tests. Maybe add a link to https://github.com/apex-dev-tools/apex-parser/issues/55 (thanks btw. to report this upstream).